staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go

```diff
@@ -83,4 +85,16 @@ type KMSConfig struct {
 	CacheSize int `json:"cachesize,omitempty"`
 	// the gRPC server listening address, for example "unix:///var/run/kms-provider.sock".
 	Endpoint string `json:"endpoint"`
+	// Timeout for gRPC calls to kms-plugin (ex. 5s).
+	// Default (5s) is defined in a const (kmsPluginConnectionTimeout) in config.go.
+	// How to select a good timeout value:
+	// 1. Check kube-apiserver.log. If you see errors originating from grpc_service.go  (in the format of
+	// "failed to create connection to unix socket"), this may imply that kms-plugin is being started after the
```

Comment:
Are these errors actually a problem or just an annoyance?

If they only occur at start up, is changing the timeout for every request really a good resolution?

Reply:
yes, my main concern is the start-up, and I am considering other options to solve this issue (see below). However, I feel that cluster deployers should have this nob.

I am working on allowing users to encrypt and decrypt existing clusters.
I have been running consecutive e2e tests for these scenarios, and I observe intermittent failures (approximately 1 in 6). In all circumstances, the failure occurred when kube-apiserver comes-up before kms-plugin and the 3 sec time-out was insufficient for those two pods to connect. Since the kube-apiserver still declares itself to be healthy even when this happens (this is what we want), other cluster components attempt to access secrets and fail - eventually leading to the failure of the kube-apiserver itself. 

Since addition of encryption and decryption occur only during cluster [setup](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/configure-helper.sh#L1906), when such disconnects happen encryption or decryption operations may fail. 
Bumping the time-out to 5 seconds solved the problem in my environment. My point is that the time-out should be exposed to the cluster deployers since they know best what their environment is like. For example, I know that kube-apiserver docker image is pre-staged on the VM, but the kms-plugin image needs to be pulled from gcr.

Note: I realize that changing the cluster bootstrapping logic may be a more direct way of solving this issue. For example, the bootstrapping script could wait for the kms-plugin to come-up before allowing kube-apiserver to start. I am considering this approach as well.